### ISSUE TITLE
Fixes and Improvements

### ISSUE THREAD

**Issue by 4512369781** [smilz0/Left4Bots#100]

**Comment by 4512369781** [comment]
> Thanks man. That is a lot of stuff, it'll take some time to test. Btw what is the bug of the bots not shooting? I never noticed it.

I spawned infected near the bots and found that they would randomly stop shooting, which reminded me of someone said bots just watched and didn't shoot to save him from charger. Finally, I sure that this bug is happen when "manual attack" is keep working, and the bots will not shoot again until there are no infected around and "manual attack" stops. So I adjust the way to set "m_isHoldingFireButton" to fix this problem.

**Comment by Mystik-Spiral** [comment]
I did not have enough time to do a proper test, so this is just feedback on my quick testing after adding your file to the VPK.  I did not notice any functional difference.  I did not notice any situations where the bots stop shooting, but I have occasionally noticed this in the past.  However, I found that was mostly, but not entirely, due to a conflict with a SourceMod plugin (Survivor Bot AI Improver).  This setting in particular:

// Makes survivor bots force attack pinned survivor's SI if possible. <0: Disabled, 1: Shoot at attacker, 2: Shove the attacker if close enough. Add numbers together.>
// -
// Default: "3"
// Minimum: "0.000000"
// Maximum: "3.000000"
ib_help_pinned_enabled "1"

When set to to 2 or 3 they would sometimes do nothing (or try shoving Chargers), so I set it to 1 (always shoot SI with pinned survivor) and it mostly resolved that issue.

With that said, I believe the "no shooting" problem happens mostly with tanks.  The bot has a weapon with ammo, but they will sometimes stop shooting and equip their melee weapon (but not attack with it).  The bots will often just stand still (hide) when the tank is not visible, even if the tank is nearby and attacking a survivor, like around a corner... they will not seek out the tank to attack it.

For proper testing I would need to disable all of my SourceMod plugins to ensure there are no issues being caused due to conflicts.  However, I will be traveling a LOT this month and will not be able to do any further testing.  I will check back after the holidays to see what the current status is.

Thanks to everyone that has created or contributed improvements to this project!

**Comment by 4512369781** [comment]
@smilz0 
Advanced
manual_attack_dual_pistol_nerf = 1
manual_attack_wandering = 0
manual_attack_common_head_radius = 300
manual_attack_special_head_radius = 300

Expert
manual_attack_dual_pistol_nerf = 0
manual_attack_wandering = 1
manual_attack_common_head_radius = 450
manual_attack_special_head_radius = 400

At higher difficulties, common infected are more dangerous, so I set it farther away.

**Comment by 4512369781** [comment]
@Mystik-Spiral 
This is mainly a bug fix, so it's not much different than before.
I don't remember seeing bots switch to melee when fighting a tank alone, but they will do so when there are common infected around. The other issue may caused by "tank_retreat_radius" settings, once the bots get close, they are ordered to retreat, so they always keep a safe distance and cannot attack when they hold a melee.
Thanks for your feedback!

**Comment by 4512369781** [comment]
@smilz0 
I updated some stuff, feel free to test it.

**Comment by 4512369781** [comment]
> This is very nice, especially the smoother aim. Good job man. I did a few tests, i haven't seen any bug. When i can i will merge the changes and update the "out" vpk.

Glad you like it! 
I wanted to do this when I first made the aiming feature, but there has some problems were not solved until now, and the addition of tolerance also makes the bots less accurate, totally it will be weaker than before. But at full power settings, two bots finally escape the Dead Center on expert difficulty after two rounds restarts, I think is good enough.

**Comment by 4512369781** [comment]
> I updated the dev vpk. Let me know if i messed up something in the process.

No, no issues found, but I updated some things again these days, please check it out.

> Update since the first version: [fixandimprove_v3.zip](https://github.com/user-attachments/files/18214486/fixandimprove_v3.zip)
> 
> 1. New setting: manual_attack_saccade_speed
> 2. Minor adjustments
> 3. Optimize find common infected
> 4. Reset bot when incapacitated and ledge

**Comment by 4512369781** [comment]
> Updated. Did i miss something? [7059fa6](https://github.com/smilz0/Left4Bots/commit/7059fa6d847005bd40f171607a25ace58ea8d0eb)

Happy New Year!
I'm working on making the rotation smoother and closer to the valve effect, so I'll update it when I'm done, and next time it should be the final version.

**Comment by 4512369781** [comment]
@smilz0 
I think this is the final version, much smoother than previous versions, and more fluid when switching targets. Also added some extra measures to prevent bot not shooting because of changed the "m_releasedFireButton" prop, and fixed the problem of bots ignoring tank rock.
The default value for "manual_attack_saccade_speed" is still set to 500, which is fine for most situations, except that it can't track fast-moving things like charger. Set it to 0 will return to the old behavior.
[fixandimpro_final.zip](https://github.com/user-attachments/files/18436720/fixandimpro_final.zip)
